typing/global_module.ml

```diff
@@ -307,3 +313,111 @@ let has_arguments t =
+
+  let rec equal_visible_parts glob1 glob2 =
+    let equal_arg_lists =
+      List.equal (Argument.equal Name.equal equal_visible_parts)
+    in
+    let equal_visible_args =
+      equal_arg_lists glob1.visible_args glob2.visible_args
+    in
+    String.equal glob1.head glob2.head && equal_visible_args
+
+  type inconsistent = Inconsistent
+
+  let meet_approximate glob1 glob2 : (t, inconsistent) Result.t =
+    match equal_visible_parts glob1 glob2 with
+    | false ->
+        Error Inconsistent
+    | true ->
+        (* Compute the meet, assuming the visible parts are equal *)
+        let rec meet glob1 glob2 =
+          let visible_args_rev =
+            Misc.Stdlib.List.merge_fold glob1.visible_args glob2.visible_args
+              ~cmp:compare_arg_name
+              ~init:[]
+              ~left_only:(fun _ _ -> assert false)
+              ~right_only:(fun _ _ -> assert false)
+              ~both:(fun acc_rev arg1 arg2 -> meet_args arg1 arg2 :: acc_rev)
+          in
+          let hidden_args_rev =
+            (* Keep only the hidden arguments that appear in both lists *)
+            Misc.Stdlib.List.merge_fold glob1.hidden_args glob2.hidden_args
+              ~cmp:compare_arg_name
+              ~init:[]
+              ~left_only:(fun acc_rev _ -> acc_rev)
+              ~right_only:(fun acc_rev _ -> acc_rev)
+              ~both:(fun acc_rev arg1 arg2 -> meet_args arg1 arg2 :: acc_rev)
+          in
+          assert (String.equal glob1.head glob2.head);
+          let visible_args = List.rev visible_args_rev in
+          let hidden_args = List.rev hidden_args_rev in
+          create_exn glob1.head visible_args ~hidden_args
+        and meet_args (arg1 : _ Argument.t) (arg2 : _ Argument.t) =
+          assert (Name.equal arg1.param arg2.param);
+          let value = meet arg1.value arg2.value in
+          ({ param = arg1.param; value } : _ Argument.t)
+        in
+        Ok (meet glob1 glob2, Approximate)
+
+  (* True if the exact name has a subset (possibly equal) of the
+     approximation's hidden arguments *)
+  let approximates ~exact ~approx =
+    let exception No in
+    equal_visible_parts exact approx
+    &&
+    match
+      Misc.Stdlib.List.merge_iter exact.hidden_args approx.hidden_args
+        ~cmp:compare_arg_name
+        ~left_only:(fun _ -> raise No)
+        ~right_only:(fun _ -> ())
+        ~both:(fun _ _ -> ())
+    with
+    | () -> true
+    | exception No -> false
+
+  let meet (t1 : t) (t2 : t) : (t, _) Result.t =
+    match t1, t2 with
+    | (glob1, Approximate), (glob2, Approximate) ->
+        meet_approximate glob1 glob2
+    | (glob1, Exact), (glob2, Exact) ->
+        begin match equal glob1 glob2 with
+        | true -> Ok t1
+        | false -> Error Inconsistent
+        end
+    | ((exact, Exact) as t_exact), (approx, Approximate)
+    | (approx, Approximate), ((exact, Exact) as t_exact) ->
+        begin match approximates ~exact ~approx with
```

Comment:
Just for the sake of reusing code, maybe we can `meet_approximate exact approx` and check that the result is equal to `exact`.

Also, maybe this function can simply raise internal error, instead of returning `Error Inconsistent`.